File: Step2/Exercise2.py
"""
Exercise2
"""

def FilterFoo(OutVar: list, *args):
    ARGN: list = args
    for item in ARGN:
        if 'Foo' in item:
            OutVar.append(item)


if __name__ == '__main__':
    # 输入列表（和 CMake 测试用例完全一样）
    InList = ["FooBar", "BarBaz", "FooBaz", "BazBar", "QuxFoo", "BazQux"]
    
    # 输出列表
    OutList = []
    
    # 调用你的 FilterFoo
    FilterFoo(OutList, *InList)
    
    # 判断是否未实现
    if not OutList:
        print("FilterFoo unimplemented or does nothing")
    
    # 测试结果标记
    Failed = False
    
    # 检查必须包含的项
    if "FooBar" not in OutList:
        print("WARNING: OutList does not contain: FooBar")
        Failed = True
    if "FooBaz" not in OutList:
        print("WARNING: OutList does not contain: FooBaz")
        Failed = True
    if "QuxFoo" not in OutList:
        print("WARNING: OutList does not contain: QuxFoo")
        Failed = True
    
    # 检查是否包含非 Foo 项
    extra_items = [item for item in OutList if "Foo" not in item]
    if extra_items:
        print(f"WARNING: OutList contains extra item(s): {extra_items}")
        Failed = True
    
    # 最终成功
    if not Failed:
        print("Success!")

File: Step2/Exercise2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

template <typename... Args> // 类似于装饰器
void FilterFoo(vector<string> &OutVar, Args... ARGN); // 变长参数ARGN


int main()
{
    bool Failed = false;
    return 0;
}


/**
 * Solution
 */

template <typename... Args> // 变长参数类型Args
void FilterFoo(vector<string> &OutVar, Args... args) // 将含有字符串"Foo"的参数追加到OutVar数组
{
    vector<string> ARGN = {args...};
    for (const string &item : ARGN)
    {
        // npos表示查找失败
        if (item.find("Foo") != string::npos)
            OutVar.push_back(item);
    }
}

/**
 * Test
 */

void check_contains(const string &var, const vector<string> &OutList, bool &Failed)
{
    auto it = find(OutList.begin(), OutList.end(), var);
    // end()表示最后一个元素的下一位，指针指向此处表示查找失败
    if (it == OutList.end())
    {
        cout << "OutList does not contain: " << var << '\n';
        Failed = true;
    }
}

// void check_nonfoo()